### PR TITLE
Pynapple type casting

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
## Content
This PR introduces a new module designed to facilitate seamless type casting between different numerical libraries: JAX, NumPy, and the pynapple library for time series data. The primary motivation behind this module is to abstract away the complexities associated with type conversions. 

This module aims to significantly reduce boilerplate code associated with type casting and data preparation.

**Typical use case:**
Calling `nemos.glm.GLM.predict` passing a `pynapple.TsdTensor`, will **now** output a `TsdTensor`, **before** it would have returned a` jax.ndarray`, dropping the time axis. Reverting back to pynapple is very cumbersome on the user side.

## Key Features:
- **Array-like Detection**: Functions to determine if objects possess array-like properties, enabling compatibility checks before performing operations.
- **Pynapple TSD Verification**: Dedicated functionality to verify if an object is a pynapple time series with data (TSD) object, ensuring correct handling of time series data.
- **Time Axis Consistency Checks**: Utilities to check for matching time axes and support intervals among TSD objects, crucial for operations involving multiple time series datasets.
- **Conditional JAX Conversion**: A mechanism to conditionally convert objects to JAX arrays if they meet specified criteria.
- **`cast_jax` Decorator**: The centerpiece of the module, this decorator automatically casts inputs to JAX arrays for processing and, where necessary, converts outputs back to pynapple TSD objects while ensuring the consistency of the time axis.

### CODE REVIEW

File: src/nemos/type_casting.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    -------
+    :
+        The original object or its conversion to a JAX array, based on the condition.
+    """
+    if condition(x):
+        x = jnp.asarray(x)
+    return x
+
+
+def cast_jax(func: Callable) -> Callable:
```
Comment (billbrod) [inline src/nemos/type_casting.py:281]: should this be "cast_jax", what we're actually doing is adding pynapple support. so we should namecheck pynapple. maybe "cast_jax_pynapple" or "support_pynapple"?

File: src/nemos/type_casting.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    """
+    flat_tree, _ = jax.tree_util.tree_flatten((args, kwargs))
+    if len(flat_tree) == 1:
+        return True
+
+    # get first pynapple
+    is_nap = list(is_pynapple_tsd(x) for x in flat_tree)
+    first_nap = is_nap.index(True)
+
+    # check time samples are close (10**-9 is hard-coded in pynapple)
```
[the remark this answers is not in the document]
Comment (billbrod) [reply src/nemos/type_casting.py:121]: or, since it sounds like pynapple doesn't allow you to do "horizontal concatenation" to combine two indepedent `Tsd` objects, and thus doesn't need to check whether time series are identical, maybe we just use the relevant float precision?

**Review by billbrod — APPROVED** [review]
This looks good to me. I have some minor thoughts that I think will be easy to address, then there are two bigger questions:
- how to type-hintg this and communicate it to user? currently, docstring and type-hinting makes it seem  like `GLM.predict` doesn't work with pynapple objects. what's the best way to communicate this behavior? Might just be: keep type-hinting as is, but describe in docstring? This is a bigger issue, and I'll create an issue for it -- doesn't need to be resolved in this PR
- I think hard-coding `10**-9` feels like it will come back to bite us. if we can't use a helper function from pynapple, then maybe we should just precision based on input dtype?

File: src/nemos/type_casting.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gviejo) [reply src/nemos/type_casting.py:121]: Pynapple doesn't have a function to check the equivalence of time series. But I can add a parameter in the config file for float precision.

[the remark this answers is not in the document]
Comment (gviejo) [reply src/nemos/type_casting.py:121]: Ok you can get it with `nap.config.nap_config.time_index_precision` and it's immutable

File: src/nemos/type_casting.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    possible array-like types or objects that mimic these properties without being suitable for
+    numerical operations.
+
+    """
+    # Check for array-like attributes
+    has_shape = hasattr(obj, "shape")
+    has_dtype = hasattr(obj, "dtype")
+    has_ndim = hasattr(obj, "ndim")
+
+    # Check for indexability (try to access the first element)
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/type_casting.py:57]: yes that makes sense

File: src/nemos/type_casting.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/type_casting.py:121]: Until realese I am installing pynapple form the main in github, @gviejo @billbrod, we should remember that at the next pynapple release we should revert that

File: src/nemos/type_casting.py (hunk 1 above)
Comment (BalzaniEdoardo) [reply src/nemos/type_casting.py:281]: I went for support_pynapple, cause it is the most explicit, I decorated the basis evaluate as well and added tests

**Comment by BalzaniEdoardo** [comment]
## Extra additions

- Decorated the basis evaluate method (if i have a time series of position in pynapple, i want to get a time series of features out. This is important If we want to do something like
```python
# predict using a time series of features and get a time series of rates
Tsd = GLM.predict(basis.evaluate(TsdFrame))
```
- Extra machinery to concatenate pynapple on axis that are not the sample (axis=0). pynapple doesn't allow it but in nemos we need to concatenate, see for example in tutorial 05 on pytrees where we use two time series as predictors.,
- Move some of the utils in tree_utils. This became necessary to aviod circular imports. The oncatenate function naturally belongs in the utils, however, utils are imported in type_casting.
